From f3e41488bd8a5dc4c3babbe94d0b131fa0ce2898 Mon Sep 17 00:00:00 2001
From: Ma <b55507@freescale.com>
Date: Mon, 4 Jan 2016 09:21:21 +0800
Subject: [PATCH 3/3] MA-7409 Better to pop up dialog to grant the permission
 when open app for the first time

Modify the SoundRecorder.java to grant the permission

Signed-off-by: Ma Jiamin<b55507@freescale.com>
---
 src/com/android/soundrecorder/SoundRecorder.java | 62 +++++++++++++++++++++++-
 1 file changed, 60 insertions(+), 2 deletions(-)

diff --git a/src/com/android/soundrecorder/SoundRecorder.java b/src/com/android/soundrecorder/SoundRecorder.java
index fba86f7..379ed8f 100644
--- a/src/com/android/soundrecorder/SoundRecorder.java
+++ b/src/com/android/soundrecorder/SoundRecorder.java
@@ -1,6 +1,6 @@
 /*
  * Copyright (C) 2011 The Android Open Source Project
- * Copyright (C) 2012 Freescale Semiconductor, Inc.
+ * Copyright (C) 2016 Freescale Semiconductor, Inc.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -21,6 +21,9 @@ import java.io.File;
 import java.text.SimpleDateFormat;
 import java.util.Date;
 
+import android.content.pm.PackageManager;
+import android.widget.Toast;
+import android.Manifest;
 import android.app.Activity;
 import android.app.AlertDialog;
 import android.content.ContentResolver;
@@ -208,7 +211,14 @@ public class SoundRecorder extends Activity
     
     static final int BITRATE_AMR =  12200; // bits/sec
     static final int BITRATE_3GPP = 12200;
-    
+
+    private int mNumPermissionsToRequest = 0;
+    private boolean mShouldRequestSoundPermission = false;
+    private boolean mShouldRequestStoragePermission = false;
+    private int mIndexPermissionRequestSound = 0;
+    private int mIndexPermissionRequestStorage = 0;
+    private static final int PERMISSION_REQUEST_CODE = 0;
+
     WakeLock mWakeLock;
     String mRequestedType = AUDIO_ANY;
     Recorder mRecorder;
@@ -246,6 +256,7 @@ public class SoundRecorder extends Activity
     public void onCreate(Bundle icycle) {
         super.onCreate(icycle);
 
+	checkPermission();
         Intent i = getIntent();
         if (i != null) {
             String s = i.getType();
@@ -859,4 +870,51 @@ public class SoundRecorder extends Activity
                 .show();
         }
     }
+
+    private void checkPermission(){
+	if(checkSelfPermission(Manifest.permission.RECORD_AUDIO)
+			!= PackageManager.PERMISSION_GRANTED){
+		mNumPermissionsToRequest++;
+		mShouldRequestSoundPermission = true;
+	}
+
+	if(checkSelfPermission(Manifest.permission.WRITE_EXTERNAL_STORAGE)
+			!= PackageManager.PERMISSION_GRANTED){
+		mNumPermissionsToRequest++;
+		mShouldRequestStoragePermission  = true;
+	}
+
+	String[] permissionToRequest = new String[mNumPermissionsToRequest];
+		int permissionRequestIndex = 0;
+	if(mShouldRequestSoundPermission){
+		permissionToRequest[permissionRequestIndex] = Manifest.permission.RECORD_AUDIO;
+		mIndexPermissionRequestSound = permissionRequestIndex;
+		permissionRequestIndex++;
+	}
+	if(mShouldRequestStoragePermission){
+		permissionToRequest[permissionRequestIndex] = Manifest.permission.WRITE_EXTERNAL_STORAGE;
+		mIndexPermissionRequestStorage= permissionRequestIndex;
+		permissionRequestIndex++;
+	}
+	if(permissionToRequest.length > 0){
+		requestPermissions(permissionToRequest, PERMISSION_REQUEST_CODE);
+	}
+    }
+
+    @Override
+    public void onRequestPermissionsResult(int requestCode,
+		String permissions[], int[] grantResults) {
+	switch (requestCode) {
+	case PERMISSION_REQUEST_CODE:
+		if (grantResults.length > 0
+				&& grantResults[0] == PackageManager.PERMISSION_GRANTED) {
+			Log.v(TAG, "Grant permission successfully");
+		} else {
+			Log.v(TAG, "Grant permission unsuccessfully");
+		}
+		break;
+	default:
+		break;
+	}
+    }
 }
-- 
1.8.0

